python/bind_operators.h

```diff
@@ -10,6 +10,41 @@
 
 namespace py = pybind11;
 
+template <class T, class... Ignored>
+void bind_common_operators(pybind11::class_<T, Ignored...> &c) {
+  c.def("__repr__", [](T &self) { return to_string(self); });
+  c.def("__bool__", [](T &) {
+    throw std::runtime_error("The truth value of a variable, data array, or "
+                             "dataset is ambiguous. Use any() or all().");
+  });
+  // For views such as VariableView __copy__ needs to return a Variable. Use
+  // result type of operator+ to obtain this type.
+  using Copy = decltype(std::declval<T>() + std::declval<T>());
+  c.def(
+      "copy", [](T &self) { return Copy(self); },
+      py::call_guard<py::gil_scoped_release>(), "Return a (deep) copy.");
+  c.def(
+      "__copy__", [](T &self) { return Copy(self); },
+      py::call_guard<py::gil_scoped_release>(), "Return a (deep) copy.");
```

Comment:
Python `copy` is a shallow copy. Is there a reason to change behaviour here? I guess keeping references for such copies would be quite messy, so probably it's better.

Reply:
Yes, the reason is a design decision taken early on in the scipp development: No sharing, or copy-on-write. I think there are several arguments for and against this. In the end I felt correctness was most important. Looking at xarray for example, their `DataArray` is essentially a dict of numpy arrays. Operations that modify data would return a copy of the array (with new data), but coords would keep pointing to the same one as the original array. This is good for performance since it can avoid copies. On the other hand, in scipp we want to do operations such as unit conversion in-place (for performance, also avoiding copies). If our coords were shared between multiple arrays this would not be possible, or break things.